codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/CodegenTarget.kt

```diff
@@ -3,7 +3,7 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-package software.amazon.smithy.rust.codegen.client.smithy.generators
+package software.amazon.smithy.rust.codegen.core.smithy
```

Comment:
I don't think core should have the concept of `CodegenTarget`. Anything that depends on should be kept in client so that it can be individually refactored into core with the target if/else cases removed.

Reply:
I agree with this sentiment in the _general case_, but:

1. Untangling the branching on `CodegenTarget` is diffcult; and
2. I'm actually not sure that removing the branching _entirely_ is beneficial. In certain _particular_ core generators, all you need is a quick and dirty `if (target == ...)` to get the job done. In some cases I think that introducing inheritance/composition/customization injections to get rid of the notion of `CodegenTarget` can be detrimental to code maintainability.

`CodegenTarget` is used by > 14 generators that this PR places in `codegen-core`. Keeping them in `codegen-client` entails bringing more things back. It's too ingrained. I think putting `CodegenTarget` in `codegen-core` is the sweet spot here.

Reply:
I opened https://github.com/awslabs/smithy-rs/issues/1765 to track this.